starky/src/verifier.rs

```diff
@@ -0,0 +1,150 @@
+    proof_with_pis: StarkProofWithPublicInputs<F, C, D>,
+    config: &StarkConfig,
+) -> Result<()>
+where
+    [(); S::COLUMNS]:,
+    [(); S::PUBLIC_INPUTS]:,
+{
+    let degree_bits = log2_strict(recover_degree(&proof_with_pis.proof, config));
+    let challenges = proof_with_pis.get_challenges(config, degree_bits)?;
+    verify_with_challenges(stark, proof_with_pis, challenges, degree_bits, config)
+}
+
+pub(crate) fn verify_with_challenges<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    S: Stark<F, D>,
+    const D: usize,
+>(
+    stark: S,
+    proof_with_pis: StarkProofWithPublicInputs<F, C, D>,
+    challenges: StarkProofChallenges<F, D>,
+    degree_bits: usize,
+    config: &StarkConfig,
+) -> Result<()>
+where
+    [(); S::COLUMNS]:,
+    [(); S::PUBLIC_INPUTS]:,
+{
+    let StarkProofWithPublicInputs {
+        proof,
+        public_inputs,
+    } = proof_with_pis;
+    let local_values = &proof.openings.local_values;
+    let next_values = &proof.openings.local_values;
+    let StarkOpeningSet {
+        local_values,
+        next_values,
+        permutation_zs,
+        quotient_polys,
+    } = &proof.openings;
+    let vars = StarkEvaluationVars {
+        local_values: &local_values.to_vec().try_into().unwrap(),
+        next_values: &next_values.to_vec().try_into().unwrap(),
+        public_inputs: &public_inputs
+            .into_iter()
+            .map(F::Extension::from_basefield)
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap(),
+    };
+
+    let (l_1, l_last) = eval_l_1_and_l_last(degree_bits, challenges.stark_zeta);
+    let last = F::primitive_root_of_unity(degree_bits).inverse();
+    let z_last = challenges.stark_zeta - last.into();
+    let mut consumer = ConstraintConsumer::<F::Extension>::new(
+        challenges
+            .stark_alphas
+            .iter()
+            .map(|&alpha| F::Extension::from_basefield(alpha))
+            .collect::<Vec<_>>(),
+        z_last,
+        l_1,
+        l_last,
+    );
+    stark.eval_ext(vars, &mut consumer);
+    let acc = consumer.accumulators();
+
+    // Check each polynomial identity, of the form `vanishing(x) = Z_H(x) quotient(x)`, at zeta.
+    let quotient_polys_zeta = &proof.openings.quotient_polys;
+    let zeta_pow_deg = challenges.stark_zeta.exp_power_of_2(degree_bits);
+    let z_h_zeta = zeta_pow_deg - F::Extension::ONE;
+    // `quotient_polys_zeta` holds `num_challenges * quotient_degree_factor` evaluations.
+    // Each chunk of `quotient_degree_factor` holds the evaluations of `t_0(zeta),...,t_{quotient_degree_factor-1}(zeta)`
+    // where the "real" quotient polynomial is `t(X) = t_0(X) + t_1(X)*X^n + t_2(X)*X^{2n} + ...`.
+    // So to reconstruct `t(zeta)` we can compute `reduce_with_powers(chunk, zeta^n)` for each
+    // `quotient_degree_factor`-sized chunk of the original evaluations.
+    for (i, chunk) in quotient_polys_zeta
+        .chunks(1 << config.fri_config.rate_bits)
```

Comment:
Do you think we should have a `quotient_degree_factor` param as in Plonky2? Looks like it's always `1 << rate_bits` currently. If we wanted to support other factors, we could have `Stark`s implement a `fn constraint_degree()` and subtract 1 from that.

I don't think it's very important to support other factors, just seems a little odd if we support it for circuits but not STARKs.

Reply:
Yeah I chose to set it to the rate by default for simplicity, but you're right that having more generality could be useful. I've added a TODO on `trait Stark`.